hcl/eval/partial.go

```diff
@@ -0,0 +1,1386 @@
+and this is obviously not possible for generating arbitrary code.
+
+The solution here involves lexing/scanning the expression bytes into tokens,
+then parsing them and evaluating the terramate symbols as they are found,
+rewriting the token stream with the primitive values, and keeping everything
+else untouched.
+*/
+
+type engine struct {
+	tokens hclwrite.Tokens
+	pos    int
+	ctx    *Context
+
+	// evalstack is a stack of evaluated nodes.
+	// The engine walks through the token list evaluating them as needed into a
+	// separated node struct placed in this stack.
+	evalstack nodestack
+
+	multiline int
+}
+
+// node represents a grammar node but in terms of its original tokens and
+// the rewritten (evaluated) ones.
+type node struct {
+	original  hclwrite.Tokens
+	evaluated hclwrite.Tokens
+
+	hasCond bool
+	hasOp   bool
+}
+
+func newPartialEvalEngine(tokens hclwrite.Tokens, ctx *Context) *engine {
+	return &engine{
+		tokens: tokens,
+		ctx:    ctx,
+	}
+}
+
+func (e *engine) Eval() (hclwrite.Tokens, error) {
+	e.newnode()
+	for e.hasTokens() {
+		err := e.evalExpr()
+		if err != nil {
+			return nil, errutil.Chain(ErrPartialEval, err)
+		}
+		e.commit()
+	}
+
+	if e.evalstack.len() != 1 {
+		panic(errorf("eval stack size must always be 1 but got %d", e.evalstack.len()))
+	}
+
+	root := e.evalstack.pop()
+	return root.evaluated, nil
+}
+
+func (e *engine) hasTokens() bool {
+	return e.pos < len(e.tokens)
+}
+
+func (e *engine) peek() *hclwrite.Token {
+	return e.tokens[e.pos]
+}
+
+func (e *engine) peekn(n int) *hclwrite.Token {
+	return e.tokens[e.pos+n]
+}
+
+// newnode creates and pushes a new node into the evaluation stack.
+func (e *engine) newnode() (int, *node) {
+	n := &node{}
+	e.evalstack.push(n)
+	return e.evalstack.len(), n
+}
+
+// commit the last node into the previous one.
+func (e *engine) commit() {
+	if e.evalstack.len() == 1 {
+		panic("everything committed")
+	}
+
+	tos := e.evalstack.pop()
+	merge := e.evalstack.pop()
+
+	merge.pushfrom(tos)
+	if tos.hasCond {
+		merge.hasCond = true
+	}
+	if tos.hasOp {
+		merge.hasOp = true
+	}
+
+	e.evalstack.push(merge)
+}
+
+func (e *engine) emit() {
+	tos := e.evalstack.peek()
+	tos.push(e.peek())
+	e.pos++
+}
+
+func (e *engine) emitn(n int) {
+	for i := 0; e.hasTokens() && i < n; i++ {
+		e.emit()
+	}
+}
+
+func (e *engine) emitVariable(v variable) {
+	tos := e.evalstack.peek()
+	tos.pushEvaluated(v.alltokens()...)
+	for i := 0; i < v.size(); i++ {
+		tos.original = append(tos.original, e.peek())
+		e.pos++
+	}
+}
+
+func (e *engine) emitTokens(original hclwrite.Tokens, evaluated hclwrite.Tokens) {
+	tos := e.evalstack.peek()
+	tos.pushEvaluated(evaluated...)
+	tos.pushOriginal(original...)
+}
+
+func (e *engine) emitnl() {
+	for e.hasTokens() && (e.peek().Type == hclsyntax.TokenNewline ||
+		e.peek().Type == hclsyntax.TokenComment) {
+		e.emit()
+	}
+}
+
+func (e *engine) emitComments() {
+	for e.hasTokens() && e.peek().Type == hclsyntax.TokenComment {
+		e.emit()
+	}
+}
```

Comment:
for the `emit` I think we will need to improve later because adding helpers() makes it even more confuse when to use each function... but thanks for the suggestion, I agree it needs work.